docs/getting-started/configuration.md
# Configuration

There're two sepparate configuration sources to setup Polygate: environment variables and the configuration file.

## Environment Variables

- `CONFIGURATION_FILE` - The path to the configuration file, this setting is required.
- `PRETTY_LOG` - Prints pretty log instead of JSON lines. The default value is "false" you can enable this feature setting it to "true".
- `ENABLE_HOT_RELOAD` - Enables configuration file change detection, when the file is changed, Polygate will stop gracefully so your orchestrator can restart it. The default value is "false" you can enable this feature setting it to "true".
- `LOG_LEVEL` - Can be: debug, info, warn or error.

## Configuration File

```yaml
redis:
  prefix: polygate # Every item on Redis will have this prefix, this avoids collisions between two applications using the same Redis
  jobPoolSize: 5 # The size of Redis connection pool to send jobs related commands like XADD, PUBLISH
  nodes: # Redis nodes to distribute the workload
    - sequence: 1 # Standalone Redis server
      host: "127.0.0.1"
      port: 6379
      db: 1 # This is not required, by default it'll pick the db 0
      password: supersecret # This is not required
    - sequence: 2 # The sequence parameter ensures the ordering of this list (it's important to the partioning algorithm)
      sentinel: true # Enables sentinel mode
      master: redismaster # Master name of sentinel
      sentinelNodes:
        - host: "172.22.1.34"
          port: 26379
        - host: "172.22.1.33"
          port: 26379
        - host: "172.22.1.32"
          port: 26379
server:
  address: "0.0.0.0" # Address to bind server's listener
  port: 4774 # Port to bind server's listener
  enable: true # Enables Polygate's producer mode
  maxHeaderListSize: 1048576 # Maximum size (bytes) of headers from a gRPC call (this parameter interferes on metadata size)
client:
  enable: true # Enables Polygate's consumer mode
metrics:
  address: "0.0.0.0" # Address to bind the metrics server
  port: 9090 # Port to bind the metrics server
  shutdownTimeout: "15s" # When Polygate is gracefully stopping this is the maximum allowed time to wait before forcing a close
  routes: # Routes must start with /
    metrics: /metrics
    readiness: /ready
    liveness: /live
protos:
  services: # Defines the accepted gRPC calls
    - service: mock.v1.MockedUnaryService # package name+"."+service name
      consumer:
        concurrency: 200 # Maximum number of jobs to be picked from Redis Streams at the same time
        block: "5000ms" # How much we must block a XREADGROUP call while waiting for jobs
        retry: # Retry policy
          limit: 3 # Maximum number of times to retry a stalled out job
          pageSize: 1000 # This is the pagination setting to iterato over the PEL list, usually you can keep this setting as 1000
          deadline: "10000ms" # How many time must be passed since a job was not acknowledged to consider stalled out
      client:
        address: "127.0.0.1" # Upstream gRPC server address
        port: 4770 # Upstream gRPC server port
      methods:
        - name: SendUnaryMock # The gRPC method name
          pattern: queue # The queue pattern will behave like request-response flow
          capped: 500 # The maximum number of messages on stream, old messages will be removed from Redis, increase this number if you have high throughput
          stream: SendUnaryMock # The Redis Stream name
    - service: mock.v1.MockedClientStreamService
      consumer:
        concurrency: 100
        block: "5000ms"
        retry:
          limit: 3
          pageSize: 1000
          deadline: "10000ms"
      client:
        address: "127.0.0.1"
        port: 4770
      methods:
        - name: SendClientStreamMock
          pattern: fireAndForget # Fire and forget pattern will behave like a pub/sub with delivery guarantees
          capped: 500
          stream: SendClientStreamMock
          timeoutWaitForNext: "300ms" # Time to consider a client stream connected to the upstream as idle to remove from consumer's pool, this timer is reset when a new job arrives
```

Create a file named `configuration.yaml` and get the IP of your Redis container:

```bash
docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' redis
```

Now just remove the Sentinel configuration as we're just using a standalone Redis and don't forget to replace the host IP. The Redis configuration will look like:

```yaml
redis:
  prefix: polygate
  jobPoolSize: 5
  nodes:
    - sequence: 1
      host: "IP_OF_YOUR_REDIS_CONTAINER"
      port: 6379
```

And do the same thing with the upstream IPs, first get the IP:

```bash
docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' grpcserver
```

Replace the IP in clients specifications:

```yaml
protos:
  services:
    - service: mock.v1.MockedUnaryService
      consumer:
        concurrency: 200
        block: "5000ms"
        retry:
          limit: 3
          pageSize: 1000
          deadline: "10000ms"
      client:
        address: "YOUR_UPSTREAM_IP" # Replace the IP here
        port: 4770
      methods:
        - name: SendUnaryMock
          pattern: queue
          capped: 500
          stream: SendUnaryMock
    - service: mock.v1.MockedClientStreamService
      consumer:
        concurrency: 100
        block: "5000ms"
        retry:
          limit: 3
          pageSize: 1000
          deadline: "10000ms"
      client:
        address: "YOUR_UPSTREAM_IP" # Replace the IP here
        port: 4770
      methods:
        - name: SendClientStreamMock
          pattern: fireAndForget
          capped: 500
          stream: SendClientStreamMock
          timeoutWaitForNext: "300ms"
```


docs/getting-started/environment-setup.md
# Environment Setup

You'll need a Redis 5+ container running:

```bash
docker run --rm --name=redis -p 6379:6379 redis:5
```

And we're going to create a mocked gRPC server, to do this you need to download the [mock](https://github.com/gamaops/polygate/tree/master/examples/mock) path from Polygate's repository. Once you downloaded the folder, navigate into it and start the gRPC mock server:

```bash
docker run --rm --name=grpcserver -p 4770:4770 -p 4771:4771 -v $(pwd):/proto tkpd/gripmock /proto/mock.proto
```

And to add the stubs execute the `setup.sh` script that is in the **mock** folder.

Now you need to create the configuration file to tell Polygate how to behave, click [here](getting-started/configuration) to learn more about Polygate's configuration file and environment variables.

CHANGELOG.md
## 2019-10-16T08:26:37.205Z
### 🚥 - adding tests (project)

#### #14

- Add pool.go unit tests

-----------------------------

## 2019-10-15T22:36:59.195Z
### 🚥 - adding tests (project)

#### #14

- Add metrics.go unit tests

-----------------------------

## 2019-10-15T11:52:18.789Z
### 🚥 - adding tests (project)

#### #14

- Add unit test to timer.go

-----------------------------

## 2019-10-15T09:45:24.279Z
### 📝 - writing docs (project)

#### #7

- Add future section

-----------------------------

## 2019-10-15T08:55:58.077Z
### 📝 - writing docs (project)

#### #7

- Add architecture scenarios

-----------------------------

## 2019-10-15T08:42:38.861Z
### 📝 - writing docs (project)

#### #7

- Add architecture overview

-----------------------------

## 2019-10-15T08:01:54.810Z
### 🆕 - add code or file(s) (project)

#### none

- Use unbuffered channel for pool to avoid keeping at least one client stream open

-----------------------------

## 2019-10-14T21:20:44.730Z
### 📝 - writing docs (project)

#### #7

- Finish the getting started section

-----------------------------

## 2019-10-14T20:52:35.119Z
### 📝 - writing docs (project)

#### #7

- Add mocked gRPC server instructions to docs

-----------------------------

## 2019-10-14T10:56:47.291Z
### 📝 - writing docs (integration)

#### #7

- Add gitter link to doc

-----------------------------

## 2019-10-14T09:31:00.787Z
### 📝 - writing docs (project)

#### #7

- Add configuration specs

-----------------------------

## 2019-10-14T08:55:19.995Z
### 🆕 - add code or file(s) (project)

#### #8

- Add metrics to count number of listeners waiting for response

-----------------------------

## 2019-10-13T22:14:50.466Z
### 📝 - writing docs (project)

#### #7

- Add favicon

-----------------------------

## 2019-10-13T22:03:01.501Z
### 🖊️ - fixing typos (project)

#### #7

- Fix doc typo

-----------------------------

## 2019-10-13T21:45:40.268Z
### 📝 - writing docs (project)

#### #7

- Add Knowledge section of docs

-----------------------------

## 2019-10-13T14:39:59.017Z
### 🆕 - add code or file(s) (project)

#### #6

- Add Prometheus metrics server and liveness/readiness probes routes

-----------------------------

## 2019-10-13T09:26:16.036Z
### 🆕 - add code or file(s) (project)

#### #5

- Import net/http/pprof to enable profiling

-----------------------------

## 2019-10-13T09:13:42.179Z
### 🆕 - add code or file(s) (project)

#### #4

- Test sentinel support and add specialized logger

-----------------------------

## 2019-10-12T22:52:59.446Z
### 🆕 - add code or file(s) (project)

#### #4

- Add Redis Sentinel support, must be tested

-----------------------------

## 2019-10-12T22:26:24.934Z
### 🆕 - add code or file(s) (project)

#### #3

- Add default values for configuration

-----------------------------

## 2019-10-12T21:52:44.067Z
### 🐛 - fixing a bug (project)

#### none

- Fix a small bug on XRETRY script (claimed capped messages should be ignored/ack)

-----------------------------

## 2019-10-12T15:47:45.878Z
### 🆕 - add code or file(s) (project)

#### #2

- Add hot reload feature

-----------------------------

## 2019-10-12T15:27:04.927Z
### 🆕 - add code or file(s) (project)

#### #1

- Add graceful stop to jobs and consumers Redis connections

-----------------------------

## 2019-10-12T15:20:28.243Z
### 🆕 - add code or file(s) (project)

#### #1

- Add graceful stop to consumers

-----------------------------

## 2019-10-12T14:31:50.337Z
### 🆕 - add code or file(s) (project)

#### none

- Finish the PoC to client stream (fire and forget pattern)

-----------------------------

## 2019-10-11T18:56:52.891Z
### 🆕 - add code or file(s) (project)

#### none

- Dynamic increase the size of conns to call ids (consumes so much memory)

-----------------------------

## 2019-10-09T15:26:43.590Z
### 🎉 - initial commit (project)

#### none

- Add initial PoC

-----------------------------



configuration.go
package main

import (
	"io/ioutil"
	"os"

	"gopkg.in/yaml.v2"

	"github.com/sirupsen/logrus"
)

type Parameters struct {
	enableHotReload   bool
	logLevel          logrus.Level
	configurationFile string
}

func loadParameters() Parameters {

	parameters := Parameters{
		enableHotReload:   true,
		logLevel:          logrus.DebugLevel,
		configurationFile: "",
	}

	configurationFile, exists := os.LookupEnv("CONFIGURATION_FILE")

	if !exists {
		log.Fatal("You must specify the CONFIGURATION_FILE environment variable")
	}

	parameters.configurationFile = configurationFile

	enableHotReload, exists := os.LookupEnv("ENABLE_HOT_RELOAD")

	if exists && enableHotReload == "false" {
		parameters.enableHotReload = false
	}

	logLevel, exists := os.LookupEnv("LOG_LEVEL")

	if exists {
		switch logLevel {
		case "info":
			parameters.logLevel = logrus.InfoLevel
		case "warn":
			parameters.logLevel = logrus.WarnLevel
		case "error":
			parameters.logLevel = logrus.ErrorLevel
		}
	}

	return parameters

}

type ConfigurationMethodExpose struct {
	Pattern            string
	Name               string
	Capped             uint64
	Stream             string
	TimeoutWaitForNext string `yaml:"timeoutWaitForNext"`
}

type ConfigurationServiceExpose struct {
	Service  string
	Consumer struct {
		Concurrency uint32
		Block       string
		Retry       struct {
			Limit    int64
			PageSize int64 `yaml:"pageSize"`
			Deadline string
		}
	}
	Client struct {
		Address string
		Port    uint16
	}
	Methods []ConfigurationMethodExpose
}

type Configuration struct {
	Redis struct {
		Prefix      string
		JobPoolSize int `yaml:"jobPoolSize"`
		Nodes       []struct {
			Sequence      uint16
			Host          string
			Port          uint16
			Db            uint8
			Password      string
			Sentinel      bool
			Master        string
			SentinelNodes []struct {
				Host string
				Port uint16
			} `yaml:"sentinelNodes"`
		}
	}
	Server struct {
		Address           string
		Port              uint16
		Enable            bool
		MaxHeaderListSize uint32 `yaml:"maxHeaderListSize"`
	}
	Client struct {
		Enable bool
	}
	Metrics struct {
		Address         string
		ShutdownTimeout string `yaml:"shutdownTimeout"`
		Port            uint16
		Routes          struct {
			Metrics   string
			Readiness string
			Liveness  string
		}
	}
	Protos struct {
		Services []ConfigurationServiceExpose
	}
}

func loadConfiguration() Configuration {

	content, err := ioutil.ReadFile(parameters.configurationFile)

	if err != nil {
		log.Fatalf("Error while reading configuration file: %v", err)
	}

	configuration := Configuration{}

	err = yaml.Unmarshal(content, &configuration)

	if err != nil {
		log.Fatalf("Error while parsing YAML from configuration file: %v", err)
	}

	defaultRedisValues(&configuration)
	defaultServerValues(&configuration)
	defaultMetricsValues(&configuration)
	defaultProtosValues(&configuration)

	return configuration

}

func defaultRedisValues(conf *Configuration) {

	if len(conf.Redis.Prefix) == 0 {
		conf.Redis.Prefix = "polygate"
	}

	if conf.Redis.JobPoolSize <= 0 {
		conf.Redis.JobPoolSize = 5
	}

	if len(conf.Redis.Nodes) == 0 {
		log.Fatalf("You must specify at least one Redis node")
	}

	for i := range conf.Redis.Nodes {

		node := &conf.Redis.Nodes[i]

		if node.Sentinel {
			if len(node.Master) == 0 {
				log.Fatalf("Master name must be specified for sentinel, node %v", i)
			}
			if len(node.SentinelNodes) == 0 {
				log.Fatalf("You must specify at least one sentinel node, node %v", i)
			}
			for t := range node.SentinelNodes {
				sentinel := &node.SentinelNodes[t]
				if len(sentinel.Host) == 0 {
					log.Fatalf("Redis Sentinel node %v must have a valid address, node %v", t, i)
				} else if sentinel.Port <= 0 {
					log.Fatalf("Redis Sentinel node %v must have a valid port, node %v", t, i)
				}
			}
			continue
		}

		if len(node.Host) == 0 {
			log.Fatalf("Redis node %v must have a valid address", i)
		} else if node.Port <= 0 {
			log.Fatalf("Redis node %v must have a valid port", i)
		}

	}

}

func defaultServerValues(conf *Configuration) {

	if len(conf.Server.Address) == 0 {
		conf.Server.Address = "0.0.0.0"
	}

	if conf.Server.Port <= 0 {
		conf.Server.Port = 4774
	}

	if conf.Server.MaxHeaderListSize <= 0 {
		log.Warnf("A maxHeaderListSize of %v bytes will disable metadata support, do you really want this?", conf.Server.MaxHeaderListSize)
	}

}

func defaultMetricsValues(conf *Configuration) {

	if len(conf.Metrics.Address) == 0 {
		conf.Metrics.Address = "0.0.0.0"
	}

	if len(conf.Metrics.ShutdownTimeout) == 0 {
		conf.Metrics.ShutdownTimeout = "15"
	}

	if conf.Metrics.Port <= 0 {
		conf.Metrics.Port = 2112
	}

	if len(conf.Metrics.Routes.Metrics) == 0 {
		conf.Metrics.Routes.Metrics = "/metrics"
	}

	if len(conf.Metrics.Routes.Readiness) == 0 {
		conf.Metrics.Routes.Readiness = "/ready"
	}

	if len(conf.Metrics.Routes.Liveness) == 0 {
		conf.Metrics.Routes.Liveness = "/live"
	}

}

func defaultProtosValues(conf *Configuration) {

	if len(conf.Protos.Services) == 0 {
		log.Fatal("You must specify at least one service")
	}

	for i := range conf.Protos.Services {

		service := &conf.Protos.Services[i]

		if len(service.Service) == 0 {
			log.Fatalf("Empty service name for service %v", i)
		}

		if service.Consumer.Concurrency <= 0 {
			service.Consumer.Concurrency = 50
		}

		if len(service.Consumer.Block) == 0 {
			service.Consumer.Block = "5000ms"
		}

		if service.Consumer.Retry.Limit <= 0 {
			service.Consumer.Retry.Limit = 3
		}

		if service.Consumer.Retry.PageSize <= 0 {
			service.Consumer.Retry.PageSize = 1000
		}

		if len(service.Consumer.Retry.Deadline) == 0 {
			service.Consumer.Retry.Deadline = "10000ms"
		}

		if len(service.Client.Address) == 0 {
			log.Fatalf("Empty client address for service %v", i)
		}

		if service.Client.Port <= 0 {
			log.Fatalf("Invalid client port for service %v", i)
		}

		if len(service.Methods) == 0 {
			log.Fatalf("You must specify at least one method for service %v", i)
		}

		for t := range service.Methods {
			method := &service.Methods[t]

			if len(method.Name) == 0 {
				log.Fatalf("Invalid method name for service %v, method %v", i, t)
			}

			if method.Pattern != "queue" && method.Pattern != "fireAndForget" {
				log.Fatalf("Invalid method pattern for service %v, method %v", i, t)
			}

			if method.Capped <= 0 {
				method.Capped = 10000
			}

			if len(method.Stream) == 0 {
				method.Stream = method.Name
			}

			if len(method.TimeoutWaitForNext) == 0 {
				method.TimeoutWaitForNext = "1s"
			}

		}

	}

}
